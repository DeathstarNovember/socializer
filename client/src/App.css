body {
  padding-top: 80px;
}

@media (max-width: 992px) {
  .container {
    width: 100% !important;
    max-width: none !important;
  }
}
