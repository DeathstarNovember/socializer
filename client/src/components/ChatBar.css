.chat-bar.sticky-top {
  top: 80px;
}

@media (max-device-width: 768px) {
  .chat-bar.sticky-top {
    height: 100%;
    padding-left: 15px;
    padding-right: 15px;
  }
}

@media (min-device-width: 768px) {
  .chat-bar.sticky-top {
    height: calc(100vh - 80px);
  }
}

.chat-conversations {
  overflow-y: scroll;
  border: 1px solid rgba(0, 0, 0, 0.125);
}

.chat,
.empty-state {
  border-top: 1px solid rgba(0, 0, 0, 0.125);
  border-bottom: 1px solid rgba(0, 0, 0, 0.125);
}

.chat:first-child {
  border-top: none;
}

.chat:last-child {
  border-bottom: none;
}

.chat:hover {
  text-decoration: none;
  color: #007bff;
  background-color: rgba(0, 0, 0, 0.05);
}

.chat:not(:last-child) {
  border-bottom: none;
}

.chat-avatars {
  flex-basis: 60px;
}

.chat-avatar-wrapper {
  width: 8px;
  overflow: visible;
}

.chat-title {
  flex: 1;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}
