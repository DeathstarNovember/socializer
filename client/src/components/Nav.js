import React, { useContext } from "react";
import { useQuery } from "@apollo/react-hooks";
import gql from "graphql-tag";
import { Container, Nav, Navbar, NavDropdown } from "react-bootstrap";
import { Link } from "react-router-dom";
import renderIf from "render-if";
import { AuthContext } from "util/context";

export const GET_USER_INFO = gql`
  query GetUserInfo {
    currentUser {
      name
    }
  }
`;

const AuthNav = () => {
  const { loading, error, data } = useQuery(GET_USER_INFO);
  const { setAuth } = useContext(AuthContext);

  if (error) {
    if (
      error.graphQLErrors[0] &&
      error.graphQLErrors[0].message === "Unauthenticated"
    ) {
      setAuth(null);
    }
    return null;
  }

  if (data && !data.currentUser) {
    return null;
  }

  return (
    <NavDropdown
      title={loading ? "User" : data.currentUser.name}
      id="profile-dropdown"
    >
      <NavDropdown.Item onClick={() => setAuth(null)}>Log out</NavDropdown.Item>
    </NavDropdown>
  );
};

const AppNav = () => {
  const { token } = useContext(AuthContext);

  return (
    <Navbar bg="primary" variant="dark" className="mb-4" fixed="top">
      <Container>
        <Link to="/" className="navbar-brand">
          Socializer
        </Link>

        {renderIf(token)(
          <Nav>
            <Link to="/" className="nav-link" role="button">
              Feed
            </Link>
            <Link to="/chat" className="nav-link" role="button">
              Chat
            </Link>
          </Nav>,
        )}

        <Navbar.Toggle />
        <Navbar.Collapse className="justify-content-end">
          <Nav>
            {renderIf(token)(<AuthNav />)}

            {renderIf(!token)(
              <>
                <Link to="/login" className="nav-link" role="button">
                  Log in
                </Link>
                <Link to="/signup" className="nav-link" role="button">
                  Sign up
                </Link>
              </>,
            )}
          </Nav>
        </Navbar.Collapse>
      </Container>
    </Navbar>
  );
};

export default AppNav;
