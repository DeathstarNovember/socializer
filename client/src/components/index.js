export { default as ChatBar } from "./ChatBar";
export { default as Conversation } from "./Conversation";
export { default as ErrorMessage } from "./ErrorMessage";
export { default as Feed } from "./Feed";
export { default as FeedItem } from "./FeedItem";
export { default as Loading } from "./Loading";
export { default as MessageThread } from "./MessageThread";
export { default as Meta } from "./Meta";
export { default as Nav } from "./Nav";
export { default as NewConversation } from "./NewConversation";
export { default as NewItem } from "./NewItem";
export { default as NewMessage } from "./NewMessage";
export { default as QueryResult } from "./QueryResult";
export { default as Posts } from "./Posts";
