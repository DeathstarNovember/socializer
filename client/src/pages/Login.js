import React, { useContext, useState } from "react";
import { useMutation } from "@apollo/react-hooks";
import { Button, Col, Container, Form, Row } from "react-bootstrap";
import { Helmet } from "react-helmet";
import gql from "graphql-tag";
import { Redirect } from "react-router-dom";
import renderIf from "render-if";
import { AuthContext } from "util/context";

export const LOGIN = gql`
  mutation Login($email: String!, $password: String!) {
    authenticate(email: $email, password: $password) {
      id
      token
    }
  }
`;

const Login = () => {
  const { token, setAuth } = useContext(AuthContext);
  const [isInvalid, setIsInvalid] = useState(false);
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");
  const [login, { data, loading, error }] = useMutation(LOGIN, {
    onError: () => setIsInvalid(true),
  });

  if (data) {
    const {
      authenticate: { id, token },
    } = data;
    setAuth({ id, token });
  }

  if (token) {
    return <Redirect to="/" />;
  }

  return (
    <>
      <Helmet>
        <title>Socializer | Log in</title>
        <meta property="og:title" content="Socializer | Log in" />
      </Helmet>
      <Container>
        <Row>
          <Col md={6} xs={12}>
            <Form
              data-testid="login-form"
              onSubmit={(e) => {
                e.preventDefault();
                login({ variables: { email, password } });
              }}
            >
              <Form.Group controlId="formEmail">
                <Form.Label>Email address</Form.Label>
                <Form.Control
                  type="email"
                  placeholder="you@gmail.com"
                  value={email}
                  onChange={(e) => {
                    setEmail(e.target.value);
                    setIsInvalid(false);
                  }}
                  isInvalid={isInvalid}
                />
                {renderIf(error)(
                  <Form.Control.Feedback type="invalid">
                    Email or password is invalid
                  </Form.Control.Feedback>,
                )}
              </Form.Group>

              <Form.Group controlId="formPassword">
                <Form.Label>Password</Form.Label>
                <Form.Control
                  type="password"
                  placeholder="Password"
                  value={password}
                  onChange={(e) => {
                    setPassword(e.target.value);
                    setIsInvalid(false);
                  }}
                  isInvalid={isInvalid}
                />
              </Form.Group>

              <Button variant="primary" type="submit" disabled={loading}>
                {loading ? "Logging in..." : "Log in"}
              </Button>
            </Form>
          </Col>
        </Row>
      </Container>
    </>
  );
};

export default Login;
