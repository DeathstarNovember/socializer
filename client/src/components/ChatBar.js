import React, { useCallback, useContext } from "react";
import { useQuery } from "@apollo/react-hooks";
import gql from "graphql-tag";
import { Link } from "react-router-dom";
import { Button } from "react-bootstrap";
import Gravatar from "react-gravatar";
import produce from "immer";
import { reverse, sortBy } from "lodash";
import renderIf from "render-if";
import { NewConversation, QueryResult } from "components";
import { Subscriber } from "containers";
import { AuthContext, ChatContext } from "util/context";
import "./ChatBar.css";

export const GET_CONVERSATIONS = gql`
  query GetConversations {
    conversations {
      id
      title
      updatedAt
      users {
        id
        gravatarMd5
      }
    }
  }
`;

export const CONVERSATIONS_SUBSCRIPTION = gql`
  subscription onConversationCreated {
    conversationCreated {
      id
      title
      updatedAt
      users {
        id
        gravatarMd5
      }
    }
  }
`;

export const CONVERSATIONS_UPDATE_SUBSCRIPTION = gql`
  subscription onConversationUpdated {
    conversationUpdated {
      id
      title
      updatedAt
      users {
        id
        gravatarMd5
      }
    }
  }
`;

const ChatBar = () => {
  const { userId } = useContext(AuthContext);

  if (userId) {
    return <AuthChatBar />;
  } else {
    return <UnauthChatBar />;
  }
};

const UnauthChatBar = () => {
  return (
    <div className="p-2 empty-state">
      <h4>Chat</h4>
      <p>
        <Link to="/login">Log in</Link> to chat
      </p>
    </div>
  );
};

const AuthChatBar = () => {
  const { userId } = useContext(AuthContext);
  const { chatState, setChatState } = useContext(ChatContext);
  const { subscribeToMore, ...queryResult } = useQuery(GET_CONVERSATIONS);
  const subscribeToNew = useCallback(() => {
    subscribeToMore({
      document: CONVERSATIONS_SUBSCRIPTION,
      updateQuery: (prev, { subscriptionData }) => {
        if (!subscriptionData.data) return prev;
        const newConversation = subscriptionData.data.conversationCreated;

        // Check that we don't already have the
        // conversation stored.
        if (prev.conversations.find((c) => c.id === newConversation.id)) {
          return prev;
        }

        return produce(prev, (next) => {
          next.conversations.unshift(newConversation);
        });
      },
    });
    subscribeToMore({
      document: CONVERSATIONS_UPDATE_SUBSCRIPTION,
      updateQuery: (prev, { subscriptionData }) => {
        if (!subscriptionData.data) return prev;
        const updatedConversation = subscriptionData.data.conversationUpdated;
        return produce(prev, (next) => {
          const match = prev.conversations.findIndex(
            (c) => c.id === updatedConversation.id,
          );
          if (match !== -1) {
            next.conversations[match] = updatedConversation;
          }
        });
      },
    });
  }, []);

  return (
    <div className="chat-bar sticky-top d-flex flex-column pb-4">
      <h4 className="d-flex justify-content-between align-items-center">
        <span>{chatState === "creating" ? "New chat" : "Chat"}</span>
        {renderIf(chatState === "default")(
          <Button
            size="sm"
            variant="primary"
            onClick={() => setChatState("creating")}
          >
            New chat
          </Button>,
        )}
      </h4>

      <div className="chat-conversations d-flex flex-column">
        <QueryResult {...queryResult}>
          {({ data }) => (
            <Subscriber subscribeToNew={subscribeToNew}>
              {renderIf(chatState === "creating")(<NewConversation />)}

              {renderIf(chatState === "default")(
                reverse(sortBy(data.conversations, "updatedAt")).map(
                  (conversation) => (
                    <Link
                      key={conversation.id}
                      to={`/chat/${conversation.id}`}
                      className="d-flex align-items-center p-2 chat"
                    >
                      <div className="d-flex chat-avatars">
                        {conversation.users
                          .filter((user) => user.id !== userId)
                          .slice(0, 3)
                          .map((user) => (
                            <div key={user.id} className="chat-avatar-wrapper">
                              <Gravatar
                                md5={user.gravatarMd5}
                                className="rounded-circle"
                                size={30}
                              />
                            </div>
                          ))}
                      </div>
                      <div className="chat-title">{conversation.title}</div>
                    </Link>
                  ),
                ),
              )}
            </Subscriber>
          )}
        </QueryResult>
      </div>
    </div>
  );
};

export default ChatBar;
