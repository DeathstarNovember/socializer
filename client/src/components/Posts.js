import React, { useCallback } from "react";
import { useQuery } from "@apollo/react-hooks";
import gql from "graphql-tag";
import produce from "immer";
import { Feed, QueryResult } from "components";

export const GET_POSTS = gql`
  query GetPosts {
    posts {
      id
      body
      insertedAt
      user {
        id
        name
        gravatarMd5
      }
    }
  }
`;

export const POSTS_SUBSCRIPTION = gql`
  subscription onPostCreated {
    postCreated {
      id
      body
      insertedAt
      user {
        id
        name
        gravatarMd5
      }
    }
  }
`;

const Posts = () => {
  const { subscribeToMore, ...queryResult } = useQuery(GET_POSTS);
  const subscribeToNew = useCallback(
    () =>
      subscribeToMore({
        document: POSTS_SUBSCRIPTION,
        updateQuery: (prev, { subscriptionData }) => {
          if (!subscriptionData.data) return prev;
          const newPost = subscriptionData.data.postCreated;

          // Check that we don't already have the
          // post stored.
          if (prev.posts.find((post) => post.id === newPost.id)) {
            return prev;
          }

          return produce(prev, (next) => {
            next.posts.unshift(newPost);
          });
        },
      }),
    [],
  );

  return (
    <>
      <h4>Feed</h4>
      <QueryResult {...queryResult}>
        {({ data }) => (
          <Feed
            feedType="post"
            items={data.posts}
            subscribeToNew={subscribeToNew}
          />
        )}
      </QueryResult>
    </>
  );
};

export default Posts;
