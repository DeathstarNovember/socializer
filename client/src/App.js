import React from "react";
import { Switch, Route } from "react-router-dom";
import { Meta, Nav } from "components";
import { StateProvider } from "containers";
import { Chat, Home, Login, Post, Signup } from "pages";
import "./App.css";

const App = (props) => {
  return (
    <StateProvider {...props}>
      <Meta />
      <Nav />

      <Switch>
        <Route path="/login" component={Login} />
        <Route path="/signup" component={Signup} />
        <Route path="/posts/:id" component={Post} />
        <Route path="/chat/:id?" component={Chat} />
        <Route component={Home} />
      </Switch>
    </StateProvider>
  );
};

export default App;
