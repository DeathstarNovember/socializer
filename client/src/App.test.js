import React from "react";
import { render, fireEvent, wait } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { MockedProvider } from "@apollo/react-testing";
import tk from "timekeeper";
import { GET_CONVERSATIONS } from "components/ChatBar";
import { GET_USER_INFO } from "components/Nav";
import { GET_POSTS } from "components/Posts";
import { LOGIN } from "pages/Login";
import App from "./App";

jest.mock("containers/Subscriber", () => ({ children }) => children);

describe("App", () => {
  beforeEach(() => {
    tk.freeze("2019-04-20");
  });

  afterEach(() => {
    tk.reset();
  });

  it("successfully proceeds through user flow logging in and out", async () => {
    const mocks = [
      {
        request: {
          query: GET_POSTS,
        },
        result: {
          data: {
            posts: [
              {
                id: 1,
                body: "Thoughts",
                insertedAt: "2019-04-18T00:00:00",
                user: {
                  id: 1,
                  name: "John Smith",
                  gravatarMd5: "abc",
                },
              },
            ],
          },
        },
      },
      {
        request: {
          query: LOGIN,
          variables: { email: "john@lvh.me", password: "password" },
        },
        result: { data: { authenticate: { id: 123, token: "abc" } } },
      },
      {
        request: {
          query: GET_USER_INFO,
        },
        result: {
          data: {
            currentUser: {
              name: "John Doe",
            },
          },
        },
      },
      {
        request: {
          query: GET_POSTS,
        },
        result: {
          data: {
            posts: [
              {
                id: 1,
                body: "Thoughts",
                insertedAt: "2019-04-18T00:00:00",
                user: {
                  id: 1,
                  name: "John Smith",
                  gravatarMd5: "abc",
                },
              },
            ],
          },
        },
      },
      {
        request: {
          query: GET_CONVERSATIONS,
        },
        result: {
          data: {
            conversations: [
              {
                id: 1,
                title: "Jane Smith, John Doe",
                updatedAt: "2019-04-18T00:00:00",
                users: [
                  {
                    id: 1,
                    gravatarMd5: "abc",
                  },
                  {
                    id: 2,
                    gravatarMd5: "def",
                  },
                ],
              },
            ],
          },
        },
      },
    ];

    const {
      container,
      getByLabelText,
      getByTestId,
      getAllByText,
      getByText,
    } = render(
      <MemoryRouter>
        <MockedProvider mocks={mocks} addTypename={false}>
          <App />
        </MockedProvider>
      </MemoryRouter>,
    );
    // Log in the user.
    fireEvent.click(getAllByText("Log in")[0]);
    fireEvent.change(getByLabelText("Email address"), {
      target: { value: "john@lvh.me" },
    });
    fireEvent.change(getByLabelText("Password"), {
      target: { value: "password" },
    });
    fireEvent.submit(getByTestId("login-form"));

    // Ensure all the queries have finished.
    await wait(() => getByText("John Doe"));
    await wait(() => getByText("Thoughts"));

    // Check that the homepage renders correctly.
    expect(container).toMatchSnapshot();

    // Log back out.
    fireEvent.click(getByText("John Doe"));
    fireEvent.click(getByText("Log out"));
    await wait(() => getAllByText("Log in")[0]);

    // Check that the homepage renders correctly.
    expect(container).toMatchSnapshot();
  });
});
