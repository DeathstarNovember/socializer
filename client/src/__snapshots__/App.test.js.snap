// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App successfully proceeds through user flow logging in and out 1`] = `
<div>
  <nav
    class="mb-4 navbar navbar-expand navbar-dark bg-primary fixed-top"
  >
    <div
      class="container"
    >
      <a
        class="navbar-brand"
        href="/"
      >
        Socializer
      </a>
      <div
        class="navbar-nav"
      >
        <a
          class="nav-link"
          href="/"
          role="button"
        >
          Feed
        </a>
        <a
          class="nav-link"
          href="/chat"
          role="button"
        >
          Chat
        </a>
      </div>
      <button
        aria-label="Toggle navigation"
        class="navbar-toggler"
        type="button"
      >
        <span
          class="navbar-toggler-icon"
        />
      </button>
      <div
        class="justify-content-end navbar-collapse collapse"
      >
        <div
          class="navbar-nav"
        >
          <div
            class="dropdown nav-item"
          >
            <a
              aria-expanded="false"
              aria-haspopup="true"
              class="dropdown-toggle nav-link"
              href="#"
              role="button"
            >
              John Doe
            </a>
            <div
              aria-labelledby=""
              class="dropdown-menu"
            >
              <a
                class="dropdown-item"
                href="#"
                role="button"
              >
                Log out
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </nav>
  <div
    class="container"
  >
    <div
      class="row"
    >
      <div
        class="d-none d-md-block col-md-4"
      >
        <div
          class="chat-bar sticky-top d-flex flex-column pb-4"
        >
          <h4
            class="d-flex justify-content-between align-items-center"
          >
            <span>
              Chat
            </span>
            <button
              class="btn btn-primary btn-sm"
              type="button"
            >
              New chat
            </button>
          </h4>
          <div
            class="chat-conversations d-flex flex-column"
          >
            <a
              class="d-flex align-items-center p-2 chat"
              href="/chat/1"
            >
              <div
                class="d-flex chat-avatars"
              >
                <div
                  class="chat-avatar-wrapper"
                >
                  <img
                    alt="Gravatar for undefined"
                    class="react-gravatar rounded-circle"
                    height="30"
                    src="//www.gravatar.com/avatar/abc?d=retro&r=g&s=30"
                    srcset="//www.gravatar.com/avatar/abc?d=retro&r=g&s=60 2x"
                    width="30"
                  />
                </div>
                <div
                  class="chat-avatar-wrapper"
                >
                  <img
                    alt="Gravatar for undefined"
                    class="react-gravatar rounded-circle"
                    height="30"
                    src="//www.gravatar.com/avatar/def?d=retro&r=g&s=30"
                    srcset="//www.gravatar.com/avatar/def?d=retro&r=g&s=60 2x"
                    width="30"
                  />
                </div>
              </div>
              <div
                class="chat-title"
              >
                Jane Smith, John Doe
              </div>
            </a>
          </div>
        </div>
      </div>
      <div
        class="col-md-8 col-12"
      >
        <h4>
          Feed
        </h4>
        <div
          class="mb-4 card"
        >
          <div
            class="card-body"
          >
            <form
              class=""
            >
              <div
                class="form-group"
              >
                <textarea
                  class="form-control"
                  placeholder="What's on your mind?"
                  rows="3"
                />
              </div>
              <button
                class="btn btn-primary"
                type="submit"
              >
                Submit
              </button>
            </form>
          </div>
        </div>
        <div
          class="mb-4 card"
        >
          <div
            class="d-flex justify-content-between align-items-center card-header"
          >
            <div
              class="d-flex align-items-center"
            >
              <img
                alt="Gravatar for undefined"
                class="react-gravatar rounded-circle"
                height="30"
                src="//www.gravatar.com/avatar/abc?d=retro&r=g&s=30"
                srcset="//www.gravatar.com/avatar/abc?d=retro&r=g&s=60 2x"
                width="30"
              />
              <div
                class="ml-2"
              >
                <div
                  class="h6 mb-0"
                >
                  John Smith
                </div>
              </div>
            </div>
            <div
              class="text-muted h7"
            >
              <time
                datetime="2019-04-18T00:00:00.000Z"
                title="2019-04-18T00:00:00Z"
              >
                2 days ago
              </time>
            </div>
          </div>
          <div
            class="card-body"
          >
            <p
              class="card-text"
            >
              Thoughts
            </p>
          </div>
          <div
            class="d-flex card-footer"
          >
            <a
              class="ml-auto"
              href="/posts/1"
            >
              Discussion ≫
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`App successfully proceeds through user flow logging in and out 2`] = `
<div>
  <nav
    class="mb-4 navbar navbar-expand navbar-dark bg-primary fixed-top"
  >
    <div
      class="container"
    >
      <a
        class="navbar-brand"
        href="/"
      >
        Socializer
      </a>
      <button
        aria-label="Toggle navigation"
        class="navbar-toggler"
        type="button"
      >
        <span
          class="navbar-toggler-icon"
        />
      </button>
      <div
        class="justify-content-end navbar-collapse collapse"
      >
        <div
          class="navbar-nav"
        >
          <a
            class="nav-link"
            href="/login"
            role="button"
          >
            Log in
          </a>
          <a
            class="nav-link"
            href="/signup"
            role="button"
          >
            Sign up
          </a>
        </div>
      </div>
    </div>
  </nav>
  <div
    class="container"
  >
    <div
      class="row"
    >
      <div
        class="d-none d-md-block col-md-4"
      >
        <div
          class="p-2 empty-state"
        >
          <h4>
            Chat
          </h4>
          <p>
            <a
              href="/login"
            >
              Log in
            </a>
             to chat
          </p>
        </div>
      </div>
      <div
        class="col-md-8 col-12"
      >
        <h4>
          Feed
        </h4>
        <div
          class="mb-4 card"
        >
          <div
            class="card-body"
          >
            <div
              class="text-muted h7"
            >
              <a
                href="/login"
              >
                Log in
              </a>
               to submit a 
              post
              .
            </div>
          </div>
        </div>
        <div
          class="mb-4 card"
        >
          <div
            class="d-flex justify-content-between align-items-center card-header"
          >
            <div
              class="d-flex align-items-center"
            >
              <img
                alt="Gravatar for undefined"
                class="react-gravatar rounded-circle"
                height="30"
                src="//www.gravatar.com/avatar/abc?d=retro&r=g&s=30"
                srcset="//www.gravatar.com/avatar/abc?d=retro&r=g&s=60 2x"
                width="30"
              />
              <div
                class="ml-2"
              >
                <div
                  class="h6 mb-0"
                >
                  John Smith
                </div>
              </div>
            </div>
            <div
              class="text-muted h7"
            >
              <time
                datetime="2019-04-18T00:00:00.000Z"
                title="2019-04-18T00:00:00Z"
              >
                2 days ago
              </time>
            </div>
          </div>
          <div
            class="card-body"
          >
            <p
              class="card-text"
            >
              Thoughts
            </p>
          </div>
          <div
            class="d-flex card-footer"
          >
            <a
              class="ml-auto"
              href="/posts/1"
            >
              Discussion ≫
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
